File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;
#define X first
#define Y second
int dist[102][102];
int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, 1, 0, -1};

int solution(vector<vector<int> > maps)
{
    for(int i=0; i<maps.size(); i++) {
        for(int j=0; j<maps[i].size(); j++) {
           dist[i][j] = -1; 
        }
    }
    dist[0][0] = 0;
    queue<pair<int, int>> q;    
    q.push({0, 0}); // 0, 0 ->  0, 1 -> 0, 2 -> 0, 3 -> 0, 4-> 1, 4 -> 2, 4
    while(!q.empty()) {
        auto cur = q.front(); 
        q.pop();
        for(int dir=0; dir<4; dir++) {
            int nx = cur.X + dx[dir];
            int ny = cur.Y + dy[dir];
            if(nx < 0 || nx >= maps.size() || ny < 0 || ny >= maps[0].size()) continue;
            if(maps[nx][ny] != 1) continue;
            if(dist[nx][ny] == -1) {
                dist[nx][ny] = dist[cur.X][cur.Y]+1;
                q.push({nx, ny});
            }
        }
    }
    
    if(dist[maps.size()-1][maps[0].size()-1] == -1) {
        return -1;
    } else {
        return dist[maps.size()-1][maps[0].size()-1]+1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, -1, 0, 1};
    static int[][] dist; 
    static int n, m;
    public int solution(int[][] maps) {      
        n = maps.length;
        m = maps[0].length;
        dist = new int[n][m];
        for(int i=0; i<n; i++) {
            Arrays.fill(dist[i], -1);
        }
        
        bfs(maps, 0, 0);
        
        int answer = dist[n-1][m-1];
        return answer;
    }
    static void bfs(int[][] maps, int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        dist[x][y] = 1;
        q.add(new int[]{x, y});
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            for(int dir=0; dir<4; dir++) {
                int nx = cur[0] + dx[dir];
                int ny = cur[1] + dy[dir];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if(dist[nx][ny] == -1 && maps[nx][ny] == 1) {
                    q.add(new int[]{nx, ny});
                    dist[nx][ny] = dist[cur[0]][cur[1]] + 1;
                }
            }
        }
    }
}